Example gallery extension

jonathansick opened the pull request:
Draft PR of the example gallery extension.

Demo of the Example Gallery generated from the Astropy documentation: http://astropy-example.jsick.codes.s3-website-us-east-1.amazonaws.com/examples/index.html

jonathansick committed:
Add sphinx_astropy.ext.example subpackage

This package will gather all Sphinx extensions related to the example
gallery. There will be two main parts: a directive that marks examples,
and extensions that index and render those examples.

This boilerplate includes the standard Sphinx setup function for the
extension.

jonathansick committed:
Add ExampleMarkerDirective

This directive marks the scope of example content in the original
documentation, and lets authors add a title and tags.

Currently a pass-through directive. It parses the content of the
directive and adds it back to the document

Examples are persisted in the build environment for later
post-processing (to build the example gallery). Examples are keyed by
their unique ID (a slugified version of the title). The dict items
contains metadata and the content of the example (to later generate
standalone example pages).

The directive also collects title and tags as metadata.

jonathansick committed:
Add example extension to v1 sphinx config

This includes the sphinx_astropy.ext.example Sphinx extension by default
in Sphinx builds.

jonathansick committed:
Insert a target node with the example source

This target node lets us backlink to the example in the main
documentation. In html, the link is an id on the first element of the
example content. example IDs are unique since they're the keys of the
env.sphinx_astropy_examples dictionary in the environment.

jonathansick committed:
Set up testing with sphinx.testing.fixtures

sphinx.testing.fixtures let us build Sphinx sites from pytest and then
inspect the built site. There can be multiple test sites, each test site
is a directory in the sphinx_astropy/test/roots/ directory.

This is the same pattern that Sphinx uses for its test, so this is likely
the easiest way to test our Sphinx extensions.
http://www.sphinx-doc.org/en/master/devguide.html#unit-testing

Note I had to add the pytest_plugins line, to load the Sphinx pytest
plugin, from a new conftest.py file at the root of the project, not from
sphinx_astropy/tests/conftest.py The reason for this is outlined in
https://docs.pytest.org/en/latest/deprecations.html#pytest-plugins-in-non-top-level-conftest-files

The rest of the pytest configuration is done in
sphinx_astropy/tests/conftest.py, which is where you'd expect most
configuration to go. This configuration is largely based on Sphinx's:
https://github.com/sphinx-doc/sphinx/blob/master/tests/conftest.py

jonathansick committed:
Add an example-gallery test site

The example-marker.rst file contains several instances of the example
directive, testing different conditions (having tags, or not, and having
different types of content in the example).

jonathansick committed:
Test that example marker creates target nodes

This test generates a site in the XML format since then it's easy to
search for nodes and their attributes.

Unfortunately this test strategy doesn't work for Sphinx <1.7, because
the pytest fixtures aren't available. Thus I have pytest skip these
tests for Sphinx <1.7.  I think this is still the best way to test
sphinx extensions and will continue to be so in the future because this
is how Sphinx tests itself.

jonathansick committed:
Test example directive's persistence to app env

jonathansick committed:
Announce new example extension in change log

jonathansick committed:
Test for examples with duplicate titles

There is a new test case (test-example-gallery-duplicates) because
otherwise the SphinxError would always be raised for regular testing of
the example directive.

jonathansick committed:
Check if duplicate example is from earlier build

As the docstring comment says, I found a weird case that while enabling
the numpydoc extension in the test environment, I would get false alarms
about duplicate examples already in the build environment. These
duplicate examples came from other test functions. Somehow the
environment is being preserved across builds now that numpydoc is
activated.

To make the ExampleMarkerDirective robust against this case, it's now
making sure that the duplicate instance is from a different document and
line number before raising a SphinxError.

jonathansick committed:
Enable debug-level logging in tests

If the tests are failing, it's useful to see the debug-level logging.

Note: the '2' verbosity enables DEBUG-level logging. I can't find a
cleaner alias to this.

Also, it would be nicer to make this the default while using the sphinx
pytest mark, but I can't find an easy way to do that.

jonathansick committed:
Test an example directive in numpydoc docstring

Now the test-example-gallery root is using an autodoc+numpydoc
processing pipeline in its build configuration. This confirms that the
directive does work in a docstring as expected.

jonathansick committed:
Remove examples in env-purge-doc event

The purge_doc callback is required to remove examples from a cached
environment if a document is removed in a subsequent build. Otherwise
the examples in the cached environment from previous builds would
continue to exist in subsequent builds.

The tests simulate a env-purge-doc event and separately ensure that
purge_examples got registered as a env-purge-doc callback.

jonathansick committed:
Refactor check for existing examples

This refactoring allows _check_for_existing_example to be used outside
the ExampleMarkerDirective, like in a env-merge-info event callback.

jonathansick committed:
Add merge_examples callback for parallel builds

This env-merge-info callback handles merging sphinx_astropy_examples
from parallel build environments when Sphinx is run in parallel read
mode.

The tests run a full-scale integration-type build with Sphinx running in
parallel (-j 4).

jonathansick committed:
Distinguish example ID from ref ID

These should be separate things so that the "example ID" is the
slugified version of the title (which is unique by design). Then the ref
ID has the example-src prefix to be a unique reference ID to the
example's source location. Also adds a ref_id field to the example's
dict in the build environment.

jonathansick committed:
Refactor example ID generation to function

This way the tranlation between a title and an example ID is codified
into an API that can be used multiple places.

jonathansick committed:
Store parsed content nodes for examples

The content_node key in the example data stores a copy of the parsed
docutils nodes for the example. It turns out that it's easier to use the
parsed content here rather than parse it during the
process_pending_example_nodes() callback where a "state" is not readily
available for parsing reStructuredText.

jonathansick committed:
Detect examples by regex

The strategy behind detect_examples() is to identify examples in the
reStructuredText source before Sphinx parses them by using a regular
expression. This lets us create stubs for example pages before Sphinx
does its regular parsing.

jonathansick committed:
Add astropy_examples_dir config variable

This configuration lets us control the directory where the example
gallery is generated.

jonathansick committed:
Add the ExamplePage class

The ExamplePage class builds upon ExampleSource, but now contains the
concerns about rendering a standalone example page.

jonathansick committed:
Implement Jinja-based templates

Templates for the standalone example pages, landing pages, tag browsing
pages, and so on, can be Jinja-formatted templates. This implementation
is adapted from sphinx.ext.autosummary, which has similar needs. By
doing this, the user can customize the templates at the builder, theme,
or project level.

The render is integrated into the ExamplePage class
(ExamplePage.render), which automatically detects and uses a template
named 'astropy_example/examplepage.rst'. The extension ships with a
default implementation of the template.

The under tests demonstrate rendering a standalone example page. This
ins't hooked up to piping in the content of the example, yet, though.

jonathansick committed:
Write stubs for example pages in builder-inited

This commit puts together the work on detecting examples from source
(detect_examples) and the work on rendering stubs for standalone example
pages (ExamplePage) and run a pipeline during the builder-inited event.
This event happens early in the Sphinx build processor so that we can
create example pages before Sphinx actually begins to read and process
these pages.

This also adds a new config variable, astropy_examples_h1, which
customizes the underline character used for making titles for "h1"
headings in reStructuredText.

jonathansick committed:
Render a basic gallery landing page

The landing page is the index.rst for the example gallery. It provides a
toctree for all the examples. The LandingPage class is implemented
similarly to the ExamplePage class in that it takes page data and is
responsible for computing paths, docnames, and rendering for itself.

In the future the template for the landing page could be enhanced into a
tiled gallery view, for example.

The test verifies that the index.rst file's reStructuredText is rendered
correctly.

Since there's now a toctree, individual example pages don't need the
`:orphan:` field.

jonathansick committed:
Add TagPage for indexing examples for given tag

The TagPage is like a specialized version of the LandingPage that
indexes exaples that have a given tag.

The TagPage.generate_tag_pages constructor simultaneously makes tag
pages for the set of tags given the population of examples, and also
provides references to those tag pages with each relevant example page.

jonathansick committed:
Landing and tag pages list tags of each example

This provides a nice way to categorize examples and to provide discovery
of other tags.

This is implemented purely in the Jinja templating layer.

jonathansick committed:
Change tag_pages to a list property

This makes it easier to use from Jinja to test lengths

jonathansick committed:
Link from example pages tags and original

This demonstrates how to use Jinja templating to provide links from a
standalone example page back to the original source page and to pages
for each associated tag.

jonathansick committed:
Add ExampleContentDirective

This directive inserts parsed content for the example from the
application environment.

jonathansick committed:
Add env-before-read-docs cb to reorder docs

The source pages are read *before* standalone example pages are read to
ensure that they can be parsed into the environment and are available
to the ExampleContentDirective.

jonathansick committed:
Insert example-content directive into page

This enables the example pages to now render the content captured by the
example marker directive

jonathansick committed:
Disable parallel_read_safe setting

Since examples are read into the environment during the same build phase
as the standalone example pages are being read, we can't guarantee that
the example is in the environment unless the build phase is performed
serially; one build process might have an example that is needed to read
in the standalone example page from another build process.

jonathansick committed:
Enable intersphinx in example-gallery test case

This will let us test intersphinx link resolution.

jonathansick committed:
Add links-demo-1 page for example-gallery tests

At the moment this page introduces an intersphinx link, which we can
test to ensure that the link's path is being corrected given the example
page's path.

jonathansick committed:
Adapt pending_xref nodes to example page

pending_xref nodes are used make internal links between Sphinx pages.
Now the ExampleContentDirective processes pending_xref nodes and changes
their refdoc to match that of the example page. This is sufficient to
ensure the link works from the standalone doc page!

jonathansick committed:
Test "ref" links in examples

The "Ref links" demo tests that an example can have a ref role pointing
to, in this case, a target in the example-marker page.

jonathansick committed:
Add an example with reference targets

This shows how reference targets work when they're part of the example
content.

jonathansick committed:
Test link to header reference target

jonathansick committed:
Add example of an equation with a label

jonathansick committed:
Add test of linking to an equation

jonathansick committed:
Make reference targets absolute for doc roles

The ExampleContentDirective now converts the reference target of a doc
role to be absolute if it was relative in the original page. This
ensures the references resolves from the standalone reference page.

Includes tests with an example containing a doc role.

jonathansick committed:
Rename the intersphinx example to API

It's not actually an intersphinx example; rather its an example of an
internal link to a Python API using the :py:obj: role.

We'll make real intersphinx links next.

jonathansick committed:
Test for external ref links with intersphinx

- Add intersphinx mapping to astropy for this test-example-gallery case
- Add a ReferenceExternalHtmlParser to parse for links with a "reference
  external" class set.
- Add a test against the intersphinx-ref-link example.

jonathansick committed:
Test an API link with intersphinx inside an example

jonathansick committed:
Make ExamplePage, TagPage, LandingPage hold app

Keeping a Sphinx app instance in these classes is useful because it
gives them direct access to application state and utilities without
having to further change their API in the future.

jonathansick committed:
Add is_directive_registered utility

This lets us check if a directive is available. It will become useful
for the doctest skipping, later. It's also useful as a check of the
extension setup (the function is tested by also testing that the
example and example-content directives got set up).

Note that this function is available in Sphinx 2.0 and later, so this
version can be retired once that becomes the minimum required version.

jonathansick committed:
Insert doctest-skip-all for example pages

- has_doctestskipall is a boolean part of the Jinja template context,
  driven by the is_directive_registered function.
- Add sphinx_astropy.ext.doctest to the example site's extension list to
  test doctest ignoring.
- Update the expected output of the ExamplePage rendering in
  reStructuredText based on the presence of the doctest-skip-all
  comment.

jonathansick committed:
Adapt image nodes in examples

This is similar to how pending_xref nodes need to be adapted if they use
a relative targetref. In this case, we adapt the uri attribute of the
image node to be an absolute URI in case a relative image path is used.
Absolute URIs and external (http) URIs are ignored.

Both the relative/local and external cases are tested here.

jonathansick committed:
Test with figure directives

Figures contain image nodes, so they are already handled by that case.

jonathansick committed:
Adapt reference nodes generated by matplotlib

Matplotlib's "plot" directive uses external links to reference the code
and image files. This is because the matplotlib extension is rendering
out-of-band with the main Sphinx build, so "ref", "doc", and "download"
roles don't work.

The trick, then, is to look for external links (a reference node) and
process refuri attributes that start with ".//", and convert those into
relative URLs.

jonathansick committed:
Demonstrates that "includes" in examples works

jonathansick committed:
Test that literalinclude works in examples

jonathansick committed:
Adapt download_reference nodes for examples

A download role creates a download_reference node that needs to be
adapted so that it points to the downloadable file from the standalone
reference page. In many ways its similar to the pending_xref node
created by ref and doc roles.

jonathansick committed:
Add astropy_examples_enabled configuration

The default for this configuration is **False**. This is done so that
the extension can be enabled in the default Sphinx configuration for
packages, and those packages can begin using the example directive. They
won't get an example gallery though, which they may not be ready for
(you need to link to the gallery index page, for instance).

This configuration then lets a package opt-in to generating the example
gallery itself.

I think this will provide a smoother adoption experience.

jonathansick committed:
Move HTML parsers to bottom of test module

This gets them out of the way of the test code; perhaps it would be
easier if they were moved to a dedicated "test utilities" module in the
code base.

jonathansick committed:
Sphinx 1.7 does not support external downloads

With Sphinx <= 1.7, :download: roles with external (i.e., https:// urls)
download links don't work. This prevents these tests from running.

jonathansick committed:
Skip additional tests with Sphinx 1.7

Named equation reference links do not seem to work with Sphinx 1.7.

As well, the format for download links from the Matplotlib plot
extension with Sphinx 1.7 are different compared to more recent Sphinx
versions, so its best to skip that test since the tests would need to be
customized for Sphinx 1.7.

jonathansick committed:
Add test with an internally-defined substitution

jonathansick committed:
Copy substitution defs in ExampleMarkerDirective

Since substitution_reference nodes are resolved just _after_ the
ExampleMarkerDirective is run, substitution_reference nodes could be part
of the example content that is republished on the standalone example
page. Thus we also need to copy the substitution_definition nodes to
include them in the standalone example page.

Based on experimentation, it seems that traversing the document's nodes
and the directive content's nodes together only gets substitution
definitions that are written above or within the example directive. This
is a caveat that will need to be added to the documentation.

jonathansick committed:
Insert substitution_def nodes into example page

This commit takes the substitution_definitions field captured by the
ExampleMarkerDirective and inserts those substitution_definition nodes
into the standalone example page.

The key part of this is to ensure we call document.note_substitution_def
because these new substitution_defs weren't already parsed.

jonathansick committed:
Note footnotes, footnote refs in example pages

Normally the document.note_footnote and document.note_footnote_refs (and
their autonumbered counterparts) are called when reStructuredText is
parsed. However, since the footnotes and footnote_refs are pre-parsed in
a standalone example page, we need to manually note them. This state is
consumed by the docutils Footnotes transform.

astrofrog commented on sphinx_astropy/ext/example/__init__.py:
```diff
@@ -0,0 +1,55 @@
+            # env_version needs to be incremented when the persisted data
+            # formats of the extension change.
+            'env_version': 1,
+            'parallel_read_safe': False,
```
@jonathansick - could you investigate what is needed to make this extension parallel-safe? Currently we are just at the point where all extensions are parallel safe for the astropy docs which provides significant improvements in build time. It would be unfortunate if we had to go backwards on that. What is the blocker currently for making this parallel-safe?

jonathansick commented:
@astrofrog Yes I saw the recent work to parallelize the Sphinx build and do feel bad that this extension runs counter to that. Here's a rough sketch of how the extension works, and why I don't think it's parallel-safe in its current implementation:

1. In the [`builder-inited`](http://www.sphinx-doc.org/en/master/extdev/appapi.html#event-builder-inited) phase, the `preprocess_examples()` function scans the docs via regex to find `example` directives. Then it populates an `examples/` directory in the documentation source tree with:

   - an index page for examples
   - pages for each tag
   - individual pages for each example

   The example pages are templated with essentially just a directive called `example-content`.

2. During the [`env-before-read-docs`](http://www.sphinx-doc.org/en/master/extdev/appapi.html#event-env-before-read-docs) phase, the `reorder_example_page_read()` function reorders the documents in `BuildEnvironment.found_docs` so that the standalone example pages in the gallery are always read last.

3. In the regular reading phase, when Sphinx encounters the `example` directive, that directive is copies the parsed nodes and stores them in the `BuildEnvironment`. When Sphinx encounters the `example-content` directive, that directive retrieves the parsed nodes of the associated example and inserts them into the page (after necessary modifications).

Step 3 is what prevents enabling parallelism. All of the `examples` directives need to be processed and be in the merged `BuildEnvironment` _before_ the `example-content` directives can be processed.

Do you see any way around this? Ideally, it would be great if there was a way to "retry" parsing a page if the environment didn't have the right metadata (i.e., the example content) yet.

One thing I looked at was having the `example-content` directive lay down something like a `pending-example-content` node and then have that node resolved in [`doctree-resolved`](http://www.sphinx-doc.org/en/master/extdev/appapi.html#event-doctree-resolved). The problem with this is that then the example content would arrive on the page too late for references to be resolved in the [`missing-reference`](http://www.sphinx-doc.org/en/master/extdev/appapi.html#event-missing-reference) phase.

Another way to do this extension, which would be parallel safe, is to make the preprocessing step, in `builder-inited`, responsible for also capturing example content (i.e., just the raw reStructuredText, not the parsed nodes). Then it wouldn't matter what order the pages are read, or if they're read in parallel. I predict the downside of this approach, though, is that that there will be more restrictions placed on how examples are written and/or will take more work to ensure that the content can be parsed. I'm thinking of things like substitutions, links, `literalinclude` directives and `include` directives, and other instances where the current architecture lets us take advantage of Sphinx/docutils to do the parsing and apply docutils Transforms for us where you very much need to be aware of the whole document, rather than just the example snippet itself.

If you have any ideas of how to rearchitect the extension, I'm absolutely open to it. This has certainly been a tricky extension to implement because of the fact that it transplants and republishes content.

astrofrog commented:
Thanks for the detailed explanation!

> Another way to do this extension, which would be parallel safe, is to make the preprocessing step, in builder-inited, responsible for also capturing example content (i.e., just the raw reStructuredText, not the parsed nodes). Then it wouldn't matter what order the pages are read, or if they're read in parallel. I predict the downside of this approach, though, is that that there will be more restrictions placed on how examples are written and/or will take more work to ensure that the content can be parsed. I'm thinking of things like substitutions, links, literalinclude directives and include directives, and other instances where the current architecture lets us take advantage of Sphinx/docutils to do the parsing and apply docutils Transforms for us where you very much need to be aware of the whole document, rather than just the example snippet itself.

I think this approach could make sense - in fact with the current approach, don't you risk making example pages that are not actually self-sufficient because they actually depend on something else on the page (code-wise), so could be missing e.g. imports and so on? Extracting the rst from the examples actually enforces better practices by making sure that the examples are fully self-contained. Or was it explicitly decided that examples don't have to be self-contained?

Just to put this out there, if you are literally transplanting content and you don't want what I mentioned above (examples to be self-contained) then I wonder if you could even consider just operating at a much later stage e.g. once the HTML is generated and transplant the relevant HTML code?

I'll have a think about other ways to work around this!

jonathansick commented:
> in fact with the current approach, don't you risk making example pages that are not actually self-sufficient because they actually depend on something else on the page (code-wise), so could be missing e.g. imports and so on?

Yes, and no. It's worked surprisingly well. You can see what types of node post-processing needs to be done in the ExampleContentDirective code: https://github.com/jonathansick/sphinx-astropy/blob/example-gallery/sphinx_astropy/ext/example/examplepages.py#L163

For instance, a lot of examples use substitutions that are defined either in the `rst_prolog` or in a separate file imported with an `include` directive. If the directive is purely copying the rst, then it would need to look for these substitution definitions, which is a lot of extra bespoke parsing that would be fragile.

At the same time, I think we should prioritize making examples as easy to mark as possible. For instance, in the demo with astropy.io.ascii many examples take advantage of a `|read|` substitution and it wouldn't be ideal if writers had to remember not to use that substitution in the scope of an example.

> I wonder if you could even consider just operating at a much later stage e.g. once the HTML is generated and transplant the relevant HTML code?

Oh wow, I hadn't thought of that. I think that would work _really_ well. All the extension would have to do at the HTML level is probably re-write some internal URLs and re-order the heading hierarchy. And since the example pages are pre-populated, Sphinx already had the doctree so the regular Sphinx toctrees and page navigation would still work.

I think I'll try that out. Sorry for not thinking of this month ago!

jonathansick committed:
Example marker wraps example in custom node

Now the ExampleMarkerDirective wraps the generated example in a custom
container node, ExampleMarkerNode.

This node has visitors that implement this node into a <div> in HTML
with a class of astropy-example-source. This div+class marks the content
of an example so that it can be copied by the HTML postprocessor into
the standalone example pages.

NOTE: is_node_registered() is backported here from Sphinx 1.8+

This small function from is needed for tests, and is no longer needed
here if Sphinx 1.8 becomes the minimum supported version of the
extension.

jonathansick committed:
ExampleContentDirective inserts a marker node

This node, ExampleContentNode, just provides an empty <div> on the HTML
page.

The class is "astropy-example-content" and the ID is the ID of the
example (which will be used to look up the example content).

Because we're no longer adding example content into the standalone
example page as part of the regular Sphinx/docutils build, we no longer
need to access and process any of the docutils nodes related to the
original example content.

Many tests are skipped because the expected content isn't be rendered
given this change. These tests will be reactivated later when examples
are being published via an HTML postprocessor.

jonathansick committed:
Examples no longer stored in BuildEnvironment

Content nodes and metadata are no longer stored in the
`sphinx_astropy_examples` attribute of the BuildEnvironment. Now that
exmaples are copied and rendered as an HTML post-processing step, we
don't need to keep this metadata in the Sphinx build. Consequently, some
additonal hooks can also go:

- merge_examples in env-merge-info
- purge_examples in env-purge-doc
- reorder_example_page_reading in env-before-read-docs

This also means that the extension can be marked as safe for reading in
parallel.

Some tests no longer apply because the
BuildEnvironment.sphinx_astropy_examples attribute is no longer set. In
many cases, I've marked these to be skipped; eventually we'll want to
reactivate these tests in the future given the new scheme.

jonathansick committed:
Make ExampleSource.docname a real docname

It will be useful for ExampleSource.docname to be a real docname so it
can be used with the BuildEnvironment APIs that translate between
docnames and paths.

Now the example page template can use a new, separate attribute
abs_docname that is useful in the doc role.

jonathansick committed:
Cache example metadata after preprocessing

Metadata is cached to the sphinx_astropy_examples attribute of the
BuildEnvironment as part of the example gallery preprocessing step
during build-inited (recall that before this caching got added as part
of the example directive.

This is necessary so that the postprocessing step know where to find the
source for each example and the associated standalone example page.

This is now also the best way to check if there are duplicate examples
since all examples are combined at this point.

Notes on tests:

1. It seems this SphinxError is build raised in the equivalent of the
test set up with the pytest.mark.sphinx pytest extension. This means
that a full build is the best way to simulate this.

2. Because docstrings are not scanned by the preprocessor, examples
embedded in docstrings are no longer part of the example gallery. Thus
I've dropped the associated test to look for example from the
example_func docstring.

jonathansick committed:
Add ExamplePage.docname attribute

This provides the actual docname, for use with Sphinx APIs.

jonathansick committed:
Add format_example_id_to_source_ref_id

Will be used to go from example ID to the reference label for the
example's source.

jonathansick committed:
Add dependency on beautifulsoup4

This is used by the example extension to manipulate built HTML pages to
insert examples into standalone pages.

jonathansick committed:
Create standalone example pages in HTML

This implements a new approach for populating standalone example pages.
Rather than republish the examples to standalone example pages as part
of the Sphinx build, this approach operates on the HTML that Sphinx has
built.

The postprocess_examples function operates as a hook for the
build-finished Sphinx event. For every example in the cached
sphinx_astropy_examples attribute of the BuildEnvironment, this method
extracts the example from the source page, adapts any relative links and
image sources, and inserts the example's <div> into the standalone
example page.

If config.astropy_examples_enabled is False, then do not run post-processing.

jonathansick committed:
Overhaul testing for html postprocessing scheme

With the large change in sphinx_astropy.ext.example to populate examples
by post processing HTML rather than copying docutils nodes, we had to
temporarily disable many of the unit tests. This commit re-works the
tests to work with the new processing strategy.

The pytest.mark.sphinx mark doesn't work with the build-finished event
(not sure why), but this means we can't use that approach to test the
HTML output. Instead, we use the CLI-based build of the example
projects. Since a lot of test functions consume those builds, I've
turned them into session-scoped fixtures.

Now that BeautifulSoup4 is a dependency of sphinx_astropy, the tests use
BeautifulSoup to check the build output. This approach is a lot cleaner
than directly using the built-in html.parser module.

Finally, I've also organized the tests around different types:

1. Unit tests that don't depend on a Sphinx build.
2. Unit tests that use the ``sphinx`` pytest mark. These tests operate on
   the Sphinx application instance after a build and test environment
   persistence.
3. Tests that run a Sphinx build through its command-line interface and
   analyze the resulting HTML product.

jonathansick committed:
Resolve anchor links outside example's scope

A link to an anchor, like #section-id, now works like this:

1. If the ID exists within the scope of the example, then the href is
left as-is. This is what the reader expects and minimizes the disruption
of follow a link in an example to a different part of the example.

2. If the ID doesn't exist in the example, but only on the source page,
then the href is adapted to point back to the source page.

jonathansick committed:
Purge examples directory on each Sphinx build

We want to support incremental rebuilds so that only those example on
pages that were changed are re-scanned and rebuilt. However, that
involves finer-grained cache invalidation. To make the example extension
work with incremental rebuilds we're starting with the *simplest* thing,
which is to start the example gallery fresh on each build. The
sphinx_astropy_examples attribute on the BuildEnvironment was already
being reset on each build; now the example directory in the source tree
itself is deleted and re-created on each Sphinx build.

jonathansick committed:
Drop importskip on versions < Sphinx 1.7

Since Sphinx 1.7 is the minimum required version, it's no longer
necessary to avoid testing against Sphinx 1.6.

jonathansick committed:
Rename "roots" to "cases" directory

The original 'roots' terminology is based on how Sphinx organizes its
own tests. Specifically, that terminology also carries over into the
pytest.mark.sphinx marker's testroot parameter.

To match sphinx-automodapi, this commit renames the 'roots' directory to
'cases' (as in, test cases). The pytest.mark.sphinx marker still works,
but I've added an extra "casesdir" pytest fixture to refer to the cases
directory with the canonical terminology.

jonathansick committed:
Add a user documentation page for the example ext

This is a draft of "getting started" documentation for
sphinx_astropy.ext.example. Right now it isn't part of a documentation
build, but could be once a Sphinx project is set up.
